bson/codec_options.py

```diff
@@ -202,20 +198,16 @@ def __eq__(self, other: Any) -> Any:
         )
 
 
-_options_base = namedtuple(  # type: ignore
-    "CodecOptions",
-    (
-        "document_class",
-        "tz_aware",
-        "uuid_representation",
-        "unicode_decode_error_handler",
-        "tzinfo",
-        "type_registry",
-    ),
-)
+class _BaseCodecOptions(NamedTuple):
+    document_class: Type[Mapping[str, Any]]
+    tz_aware: bool
+    uuid_representation: int
+    unicode_decode_error_handler: Optional[str]
+    tzinfo: Optional[datetime.tzinfo]
+    type_registry: TypeRegistry
```

Comment:
Although I'm a little wary of this change in 3.6.1: https://docs.python.org/3/library/typing.html#typing.NamedTuple

> Changed in version 3.6.1: Added support for default values, methods, and docstrings.

Does this new typing.NamedTuple approach work on python 3.6.0?

Reply:
Spun up a RHEL 7.0 FIPS host to test tomorrow

Reply:
Wow, I can't even get `mypy` to run on a file that has `print('hello')`:

```python
Traceback (most recent call last):
  File "/root/.local/bin/mypy", line 5, in <module>
    from mypy.__main__ import console_entry
  File "/root/.local/lib/python3.6/site-packages/mypy/__main__.py", line 6, in <module>
    from mypy.main import main, process_options
  File "mypy/main.py", line 13, in <module>
  File "mypy/build.py", line 32, in <module>
  File "mypy/semanal.py", line 85, in <module>
  File "mypy/errors.py", line 14, in <module>
  File "mypy/message_registry.py", line 29, in <module>
TypeError: __new__() missing 1 required positional argument: 'code'
```

I think we should stipulate that we support typing on Python 3.7+ only.

Reply:
Oof. Can you report that error as a bug to mypy?

So mypy doesn't work, but does the driver function correctly? And does it work with my suggestion above? If not we should add a comment that _BaseCodecOptions is required for 3.6.0 support.

Reply:
It runs fine on Python 3.6.15, so I don't think there's anything to report, since Python 3.6 is EOL anyway.

Reply:
It looks like with Python 3.6.0 we can't use `CodecOptions(SON[str, Any])`.  It passes in CI but not in FIPS.  

```python
[2022/03/11 18:54:31.415] ----------------------------------------------------------------------
[2022/03/11 18:54:31.415] Traceback (most recent call last):
[2022/03/11 18:54:31.415]   File "/data/mci/5eb993a02c41b93b14bde7085d451fe9/src/test/__init__.py", line 561, in wrap
[2022/03/11 18:54:31.415]     return f(*args, **kwargs)
[2022/03/11 18:54:31.415]   File "/data/mci/5eb993a02c41b93b14bde7085d451fe9/src/test/test_binary.py", line 261, in test_legacy_csharp_uuid_roundtrip
[2022/03/11 18:54:31.415]     docs = bson.decode_all(data, CodecOptions(SON[str, Any], False, CSHARP_LEGACY))
[2022/03/11 18:54:31.415]   File "/data/mci/5eb993a02c41b93b14bde7085d451fe9/src/bson/son.py", line 69, in __new__
[2022/03/11 18:54:31.415]     instance = super(SON, cls).__new__(cls, *args, **kwargs)
[2022/03/11 18:54:31.415] TypeError: super(type, obj): obj must be an instance or subtype of type
[2022/03/11 18:54:31.415] ======================================================================
[2022/03/11 18:54:31.415] ERROR: test_legacy_java_uuid (test_binary.TestBinary)
[2022/03/11 18:54:31.415] ----------------------------------------------------------------------
```

We should be able work around this in the tests by having a version guard.

Reply:
Alternately, we could set our minimum supported Python to 3.6.2  (enforceable in `python_requires`) and bring back support for `NoReturn` and several other things that were added to typing in 3.6.1 and 3.6.2.

Reply:
> It runs fine on Python 3.6.15, so I don't think there's anything to report, since Python 3.6 is EOL anyway.

Mypy claims to support >=3.6.0 so this would be a real bug to report.

Reply:
> Mypy claims to support >=3.6.0 so this would be a real bug to report.

It feels like a waste of our time to find a repro and report this in my opinion.